conductor_api/src/keystore.rs

```diff
@@ -0,0 +1,763 @@
+    sync::{Arc, Mutex},
+};
+
+const PCHECK_HEADER_SIZE: usize = 8;
+const PCHECK_HEADER: [u8; 8] = *b"PHCCHECK";
+const PCHECK_RANDOM_SIZE: usize = 32;
+const PCHECK_SIZE: usize = PCHECK_RANDOM_SIZE + PCHECK_HEADER_SIZE;
+const KEYBUNDLE_SIGNKEY_SUFFIX: &str = ":sign_key";
+const KEYBUNDLE_ENCKEY_SUFFIX: &str = ":enc_key";
+
+pub enum Secret {
+    SigningKey(SigningKeyPair),
+    EncryptingKey(EncryptingKeyPair),
+    Seed(SecBuf),
+}
+
+enum KeyType {
+    Signing,
+    Encrypting,
+}
+
+#[allow(dead_code)]
+#[derive(Serialize, Deserialize)]
+pub struct Keystore {
+    passphrase_check: String,
+    secrets: BTreeMap<String, KeyBlob>,
+    #[serde(skip_serializing, skip_deserializing)]
+    cache: HashMap<String, Arc<Mutex<Secret>>>,
+    #[serde(skip_serializing, skip_deserializing)]
+    passphrase_manager: Option<Arc<PassphraseManager>>,
+    #[serde(skip_serializing, skip_deserializing)]
+    hash_config: Option<PwHashConfig>,
+}
+
+fn make_passphrase_check(
+    passphrase: &mut SecBuf,
+    hash_config: Option<PwHashConfig>,
+) -> HcResult<String> {
+    let mut check_buf = SecBuf::with_secure(PCHECK_SIZE);
+    check_buf.randomize();
+    check_buf.write(0, &PCHECK_HEADER).unwrap();
+    encrypt_with_passphrase_buf(&mut check_buf, passphrase, hash_config)
+}
+
+impl Keystore {
+    #[allow(dead_code)]
+    pub fn new(
+        passphrase_manager: Arc<PassphraseManager>,
+        hash_config: Option<PwHashConfig>,
+    ) -> HcResult<Self> {
+        Ok(Keystore {
+            passphrase_check: make_passphrase_check(
+                &mut passphrase_manager.get_passphrase()?,
+                hash_config.clone(),
+            )?,
+            secrets: BTreeMap::new(),
+            cache: HashMap::new(),
+            passphrase_manager: Some(passphrase_manager),
+            hash_config,
+        })
+    }
+
+    #[allow(dead_code)]
+    pub fn new_from_file(
+        path: PathBuf,
+        passphrase_manager: Arc<PassphraseManager>,
+        hash_config: Option<PwHashConfig>,
+    ) -> HcResult<Self> {
+        let mut file = File::open(path)?;
+        let mut contents = String::new();
+        file.read_to_string(&mut contents)?;
+        let mut keystore: Keystore = serde_json::from_str(&contents)?;
+        keystore.hash_config = hash_config;
+        keystore.passphrase_manager = Some(passphrase_manager);
+        Ok(keystore)
+    }
+
+    #[allow(dead_code)]
+    fn check_passphrase(&self, mut passphrase: &mut SecBuf) -> HcResult<bool> {
+        let mut decrypted_buf = decrypt_with_passphrase_buf(
+            &self.passphrase_check,
+            &mut passphrase,
+            self.hash_config.clone(),
+            PCHECK_SIZE,
+        )?;
+        let mut decrypted_header = SecBuf::with_insecure(PCHECK_HEADER_SIZE);
+        let decrypted_buf = decrypted_buf.read_lock();
+        decrypted_header.write(0, &decrypted_buf[0..PCHECK_HEADER_SIZE])?;
+        let mut expected_header = SecBuf::with_secure(PCHECK_HEADER_SIZE);
```

Comment:
I'm not sure I'm following this passphrase header bit. Perhaps some comments? also, one is in a secure buffer and one is in an insecure buffer. If they are expected to be the same data, probably should have the same security.

Reply:
So, we are encrypting `PCHECK_HEADER` followed by 32 random bytes with the passphrase and are storing this cipher in `Keystore::passphrase_check`. Everytime the keystore gets modified, we have to encrypt the new secret but need to make sure it is encrypted with the same passphrase. Before using the provided passphrase to encrypt the new secret, we try to decrypt `passphrase_check` and expect to get back our `PCHECK_HEADER` followed by some random bytes. We only compare `PCHECK_HEADER` and ignore the random bytes during check.

Yeah, I don't think this second `SecBuf` needs to be secure..

Will add comments.

Reply:
:+1: